### PR TITLE
Add missing functions from Array to Dynarray

### PR DESCRIPTION
**Pull request by jake-87**
Added:
- mem
- memq
- find_opt
- find_index
- find_map
- find_mapi

All implementations are based off the implementation in Array. I have tried my best, based on the given comments, and other functions, to add all the correct checks; if I have missed any, please feel free to note or edit, and similar with the quality/number of tests - please inform me if these are not sufficient.

I also apologies; I could not quite figure out the format of the "Changes" file, so I haven't included that change here currently.

Background:
I was attempting to port some code from `List` to `Dynarray`; I promptly realized that `Dynarray` did not have `find_opt` or similar, and decided that it would probably be easier to implement them myself than raise an issue and wait for someone else to do it.

### CODE REVIEW

**Review by gasche — COMMENTED** [review]
This looks fine, but I would propose to systematically (and consistently) call `check_same_length` outside the `loop` function:

```
let rec loop i = ...
in
let res = loop 0 in
check_same_length "..." a ~length;
res
```

**Comment by jake-87** [comment]
Sure, that seems reasonable. I'll make that change.

**Comment by gasche** [comment]
Could you call `check_same_length` in mem, memq as well? (They don't call user-provided functions but we can still observe concurrent changes.)

**Comment by jake-87** [comment]
Sure.

**Comment by jake-87** [comment]
Hey, sorry for the bump - is there anything else you would like me to implement/change?

**Comment by jake-87** [comment]
Hey @gasche; apologies for the bump. Is there anything else you'd like me to implement/change for this patch?

**Review by gasche — APPROVED** [review]
I believe that this is correct, and it seems sensible to complete the API with more `Array` functions.

Per the rules of stdlib maintenance, this needs an approval from another maintainer before we can consider merging this. (Hopefully one of them will receive this in their mailbox and decide to have a look.)

**Comment by Octachron** [comment]
Looking at the `Array` module documentation, the new functions are part of the `Array scanning` section, which also contains `exists2` and `forall2`. However, It might make sense to exclude those n-array functions that require more synchronizations from `Dynarray`.

File: stdlib/dynarray.mli
Comment (Octachron) [inline stdlib/dynarray.mli:313]: For the documentation part, I would propose to mirror the `Array` documentation by moving `forall` and `exists`  just before `mem` and adding a
```
{1:dynarray_scanning Dynarray scanning }
```
header.

**Comment by Octachron** [comment]
Note that I think it is probably better to left the hypothetical addition of `exists2` and `for_all2` for another PR which would add the `*2` functions to  `Dynarray`.

**Comment by jake-87** [comment]
Sorry about that; must have edited the file after I ran the scripts.

**Review by Octachron — APPROVED** [review]

**Comment by gasche** [comment]
I believe that the new functions are missing `@since 5.3` tags in the documentation,

**Comment by gasche** [comment]
... and a Changes entry in the "Standard library" section.

**Comment by jake-87** [comment]
Added. Apologies if the `Changes` file addition is not correct; as mentioned above, I'm not quite sure of the format.

**Comment by gasche** [comment]
Merged, thanks!

**Comment by jake-87** [comment]
Thank you! (and thank you both for putting up with my learnings of the process 😅)

**Pull request merged**
